### PR TITLE
Assignment6 - Ian Kaiserman

### PR DESCRIPTION
**Pull request by Afterlyte**
Paired with @eroadhouse

### CODE REVIEW

**Comment by Tom-Rohr** [comment]
I'll review this.

File: Calculate/Calculator.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            return true;
+        }
+
+        public Calculator()
+        {
+            _MathematicalOperations.Add('+', Add);
+            _MathematicalOperations.Add('-', Subtract);
+            _MathematicalOperations.Add('*', Multiply);
+            _MathematicalOperations.Add('/', Divide);
+        }
```
Comment (Tom-Rohr) [inline Calculate/Calculator.cs:41]: You could get rid of this constructor and just add these with an object initializer when you initialize IReadOnlyDictionary if you didnt want to include your own dvc.

File: Calculate/Calculator.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
+﻿namespace Calculate
+{
+    public class Calculator
+    {
+        private readonly Dictionary<char, Func<int, int, int>> _MathematicalOperations = new();
+        public IReadOnlyDictionary<char, Func<int, int, int>> MathematicalOperations { get => _MathematicalOperations; }
+        public static int Add(int x, int y) { return x + y; }
+        public static int Subtract(int x, int y) { return x - y; }
+        public static int Multiply(int x, int y) { return x * y; }
+        public static int Divide(int x, int y) { return x / y; }
```
Comment (Tom-Rohr) [inline Calculate/Calculator.cs:10]: Might want to make the return type for these and your Funcs doubles. Calculator wont give a correct result if the two ints you're diving doesn't come out to a whole number.

**Review by Tom-Rohr — COMMENTED** [review]
## Instructions

- Create a *Console* project called "Calculate.". ✔
- Define a Program Class
  - Define two init-only setter properties, `WriteLine` and `ReadLine`, that contain delegates for writing a line of text and reading a line of text respectively ✔
  - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
  - Set the default behavior for the `WriteLine` and `ReadLine` properties to invoke `System.Console` versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
  - Define static `Add`, `Subtract`, `Multiple`, and `Divide` methods that have two parameters and return a third parameter. ✔
  - Define a read-only property, `MathematicalOperations`, of type `System.Collections.Generics.IReadOnlyDictionary<TKey,TValue>` that:
    - is initialized to a `System.Collections.Generics.Dictionary<<TKey,TValue>` instance that. ✔
      - Uses `char` for the key corresponding to the operators +, -, *, and /. ✔
      - Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
  - Implement a `TryCalculate` method following "TryParse" pattern ✔
    - Valid `calculation` expressions include such strings as "3 + 4", "42 - 2", etc. ❌
    - If there is no whitespace around the operator, you can assume the `calculation` is invalid and return false. Similarly if the operands are not integers. ✔
    - Use `string.Split()`, pattern matching, logical and operators to parse the string in their entirety ✔
    - Index into the `MathematicalOperations` method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the `WriteLine`/`ReadLine` properties on `Program` for testing the input and output of your program. ✔

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- Place all shared project items into a `Directory.Build.targets` file.
- nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ❌(Might want more for division)
- Choose simplicity over complexity ✔

File: Calculate/Calculator.cs (hunk 2 above)
Comment (DaisyPalencia) [inline Calculate/Calculator.cs:10]: I agree with Tom-Rohr here; I would recommend changing the return type to Double for these functions.

File: Calculate/Calculator.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
+﻿namespace Calculate
+{
+    public class Calculator
+    {
+        private readonly Dictionary<char, Func<int, int, int>> _MathematicalOperations = new();
+        public IReadOnlyDictionary<char, Func<int, int, int>> MathematicalOperations { get => _MathematicalOperations; }
+        public static int Add(int x, int y) { return x + y; }
+        public static int Subtract(int x, int y) { return x - y; }
+        public static int Multiply(int x, int y) { return x * y; }
```
Comment (DaisyPalencia) [inline Calculate/Calculator.cs:9]: I would also recommend considering dividing by zero in your divide function.

**Review by DaisyPalencia — COMMENTED** [review]
Instructions
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ❌
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

File: Calculate/Program.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public class Program
+    {
+        public Func<string?> ReadLine { get; init; } = System.Console.ReadLine;
+
+        public Action<string> WriteLine { get; init; } = System.Console.WriteLine;
+
+        public Calculator Calc { get; }
+
+        public Program()
```
Comment (mmwoodfo) [inline Calculate/Program.cs:11]: You should have another constructor here for setting the read/write line properties

File: Calculate.Tests/CalculatorTests.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+{
+    [TestClass]
+    public class CalculatorTests
+    {
+        [TestMethod]
+        public void TryCalculate_GivenExpression_ReturnsTrueResult()
+        {
+            Calculator calculator = new();
+
+            calculator.TryCalculate("1 + 2", out double result1);
```
Comment (mmwoodfo) [inline Calculate.Tests/CalculatorTests.cs:14]: This would be a great time to use data driven tests!

**Pull request closed without merging**